### PR TITLE
DM-257 CRUD resource actions for CWU proposals

### PR DESCRIPTION
**Pull request by sutherlanda**
This PR includes resource CRUD actions for CWU proposals.

### CODE REVIEW

**Comment by sutherlanda** [comment]
This is ready for review @dhruvio

File: src/back-end/lib/resources/proposal/code-with-us.ts [hunk 1]
```diff
@@ -0,0 +1,403 @@
...
+        switch (tag) {
+          case 'edit':
+            return adt('edit', {
+              proposalText: getString(value, 'proposalText'),
+              additionalComments: getString(value, 'additionalComments'),
+              proponent: get(value, 'proponent'),
+              attachments: getStringArray(value, 'attachments')
+            });
+          case 'submit':
+            return adt('submit', String(value));
```
Comment (dhruvio) [inline src/back-end/lib/resources/proposal/code-with-us.ts]: Change `String(value)` to `getString(body, 'value', '')`. If value is `undefined` or `null`, the current approach will simply store it as `"undefined"` or `"null"`.

Comment (dhruvio) [reply src/back-end/lib/resources/proposal/code-with-us.ts]: Same for the cases below.

File: src/shared/lib/resources/proposal/code-with-us.ts [hunk 2]
```diff
@@ -2,9 +2,10 @@ import { CWUOpportunitySlim } from 'shared/lib/resources/code-with-us';
 import { FileRecord } from 'shared/lib/resources/file';
 import { Organization } from 'shared/lib/resources/organization';
 import { User } from 'shared/lib/resources/user';
-import { ADT, Id } from 'shared/lib/types';
+import { ADT, BodyWithErrors, Id } from 'shared/lib/types';
+import { ErrorTypeFrom } from 'shared/lib/validation';
 
-export enum ProposalStatus {
+export enum CWUProposalStatus {
```
Comment (dhruvio) [inline src/shared/lib/resources/proposal/code-with-us.ts]: This needs another status — `Evaluated` for when a score has been entered.

File: src/shared/lib/resources/proposal/code-with-us.ts [hunk 3]
```diff
@@ -2,9 +2,10 @@ import { CWUOpportunitySlim } from 'shared/lib/resources/code-with-us';
...
   Draft = 'DRAFT',
   Submitted = 'SUBMITTED',
   Review = 'REVIEW',
```
Comment (dhruvio) [inline src/shared/lib/resources/proposal/code-with-us.ts]: Is it possible to rename this to `UnderReview`?

File: src/back-end/lib/resources/proposal/code-with-us.ts [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+            const validatedSubmissionNote = proposalValidation.validateNote(request.body.value);
+            if (isInvalid(validatedSubmissionNote)) {
+              return invalid({ proposal: adt('submit' as const, validatedSubmissionNote.value) });
+            }
+            return valid({
+              session: request.session,
+              body: adt('submit', validatedSubmissionNote.value)
+            });
+          case 'score':
+            if (!isValidStatusChange(validatedCWUProposal.value.status, CWUProposalStatus.Review)) {
```
Comment (dhruvio) [inline src/back-end/lib/resources/proposal/code-with-us.ts]: The `to` status should be `Evaluated`.

File: src/shared/lib/resources/proposal/code-with-us.ts (hunk 3 above)
Comment (dhruvio) [reply src/shared/lib/resources/proposal/code-with-us.ts]: Status changes to `UnderReview` will happen automatically when a proposal deadline for an opportunity lapses. We should talk about how to handle this. Either via a daemon, or on-demand.

File: src/back-end/lib/resources/proposal/code-with-us.ts [hunk 5]
```diff
@@ -0,0 +1,403 @@
...
+          return invalid({
+            permissions: [permissions.ERROR_MESSAGE]
+          });
+        }
+        const validatedCWUProposal = await validateCWUProposalId(connection, request.params.id, request.session);
+        if (isInvalid(validatedCWUProposal)) {
+          return invalid({ notFound: ['Proposal not found.'] });
+        }
+        if (validatedCWUProposal.value.status !== CWUProposalStatus.Draft) {
+          return invalid({ permissions: [permissions.ERROR_MESSAGE] });
```
Comment (dhruvio) [inline src/back-end/lib/resources/proposal/code-with-us.ts]: Change this error message:

`{ status: ['You can not delete a proposal that is not a draft.'] }`

File: src/back-end/lib/db.ts [hunk 6]
```diff
@@ -1080,3 +1102,457 @@ export const deleteCWUOpportunity = tryDb<[Id], CWUOpportunity>(async (connectio
...
+  return valid(await connection.transaction(async trx => {
+    // Update status for proposal first
+    const [result] = await connection<CWUProposalStatusRecord>('cwuProposalStatuses')
+      .transacting(trx)
+      .insert({
+        id: generateUuid(),
+        proposal: proposalId,
+        createdAt: now,
+        createdBy: session.user.id,
+        status: CWUProposalStatus.Review,
```
Comment (dhruvio) [inline src/back-end/lib/db.ts]: Change to `Evaluated`.

File: src/back-end/lib/validation.ts [hunk 7]
```diff
@@ -117,3 +120,44 @@ export async function validateCWUOpportunityId(connection: db.Connection, opport
...
+type ValidatedCreateProponentRequestBody = CreateProponentRequestBody;
+
+export async function validateProponent(connection: db.Connection, raw: any): Promise<Validation<ValidatedCreateProponentRequestBody, CreateProponentValidationErrors>> {
+  switch (get(raw, 'tag')) {
+    case 'individual':
+      const validatedIndividualProponentRequestBody = validateIndividualProponent(get(raw, 'value'));
+      if (isValid(validatedIndividualProponentRequestBody)) {
+        return valid(adt('individual', validatedIndividualProponentRequestBody.value));
+      }
+      return invalid(adt('individual', validatedIndividualProponentRequestBody.value));
```
[the remark this answers is not in the document]
Comment (sutherlanda) [reply src/back-end/lib/validation.ts]: Had to add a `as const` but it typechecks :)

**Comment by sutherlanda** [comment]
@dhruvio I've implemented the requested changes.  I've also added `statusHistory` for proposal responses to this branch.  `statusHistory` for opportunities is in `feature/andrew/cwu-opportunity-statuses`, but let me know if you want everything here.

**Pull request merged**